src/librustdoc/clean/mod.rs

```diff
@@ -1496,10 +1497,42 @@ impl Clean<Type> for hir::Ty<'_> {
     }
 }
 
+/// Returns `None` if the type could not be normalized
+fn normalize(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'tcx>> {
+    use crate::rustc_trait_selection::infer::TyCtxtInferExt;
+    use crate::rustc_trait_selection::traits::query::normalize::AtExt;
+    use rustc_middle::traits::ObligationCause;
+    use rustc_middle::ty::ParamEnv;
+
+    // Try to normalize `<X as Y>::T` to a type
+    // FIXME: rustdoc won't be able to perform 'partial' normalization
+    // until this param env is actually correct
+    // 'partial': `<Vec<T> as IntoIterator>::IntoIter>` -> `vec::IntoIter<T>`
+    let param_env = ParamEnv::empty();
+    let lifted = ty.lift_to_tcx(tcx).unwrap();
+    let normalized = tcx.infer_ctxt().enter(|infcx| {
+        infcx
+            .at(&ObligationCause::dummy(), param_env)
+            .normalize(lifted)
+            .map(|resolved| infcx.resolve_vars_if_possible(resolved.value))
+    });
+    match normalized {
+        Ok(normalized_value) => {
+            debug!("resolved {:?} to {:?}", ty, normalized_value);
+            Some(normalized_value)
+        }
+        Err(err) => {
+            debug!("failed to resolve {:?}: {:?}", ty, err);
+            None
+        }
+    }
+}
+
 impl<'tcx> Clean<Type> for Ty<'tcx> {
     fn clean(&self, cx: &DocContext<'_>) -> Type {
         debug!("cleaning type: {:?}", self);
-        match *self.kind() {
+        let ty = normalize(cx.tcx, self.lift_to_tcx(cx.tcx).unwrap()).unwrap_or(self);
```

Comment:
you lift the `self` (`ty` in `normalize`) in `normalize` anyway, so you can just pass it directly here.

Reply:
Thanks, it gave me trouble the first time because I made the lifetime bounds on `normalize()` too strict - it turns out I can change `Ty<'tcx>` to `Ty<'_>` and now it works.